### PR TITLE
Uncertainty model

### PR DESCRIPTION
**Pull request by Serafim179**
First version of the LSTM model with MC-dropout uncertainty estimation on four given datasets. 

The features used are:
- 3 preprocessed PPG channels as time series
- HR as static feature

### CODE REVIEW

**Comment by gianlucatruda** [comment]
> HR as static feature

@Serafim179 does this mean you're using the ground truth value for the heart rate as an input feature for the model? That kinda violates the goals of the CoVital models, as we are trying to predict SPo2 based only on the video data. Or have I misunderstood something?

File: modelling/uncertainty_model/model.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    inputs = Input(shape=(n_steps, n_channels))
+    x = LSTM(n_samples, activation='tanh', return_sequences=True)(inputs)
+    x = Dropout(0.2)(x, training=True)
+    x = LSTM(n_samples, activation='tanh')(x)
+    x = Dense(n_samples, activation='relu')(x)
+    x = Dropout(0.2)(x, training=True)
+    out = Dense(1, activation='linear')(x)
+
+    return Model(inputs=inputs, outputs=out)
```
Comment (gianlucatruda) [inline modelling/uncertainty_model/model.py:42]: This is great 👏 Would be useful to have brief docstrings though.

File: run.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+from modelling.feat_engineer.data_processing import *
+
+from keras import optimizers
+
+
+clin_X, clin_y, clin_z = make_seq('preprocessed_data/clinical_data/covital_clinical.csv', 'preprocessed_data/clinical_data/ground_truths_clinical.csv', return_features=True)
+com_X, com_y, com_z = make_seq('preprocessed_data/community_data/covital_community.csv', 'preprocessed_data/community_data/ground_truths_community.csv', return_features=True)
+sample_X, sample_y, sample_z = make_seq('preprocessed_data/sample_data/sample_data.csv', 'preprocessed_data/sample_data/ground_truths_sample.csv', return_features=True)
+nemc_X, nemc_y, nemc_z = make_seq('preprocessed_data/nemcova_data/nemcova_data.csv', 'preprocessed_data/nemcova_data/ground_truths_nemcova.csv', return_features=True)
+
```
Comment (gianlucatruda) [inline run.py:16]: We can merge this with my code, then just import the unified (and tidied) data.

File: run.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+data = [X, features]
+
+model = create_model(X, features)
+
+learning_rate = 0.01
+opt = optimizers.Adam(lr=learning_rate)
+model.compile(optimizer=opt, loss='mse')
+# fit model
+history_cache = model.fit(data, y, epochs=150, verbose=1, batch_size=29)
+model.save('model.h5')
```
Comment (gianlucatruda) [inline run.py:29]: Nice!

File: test.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    else:
+        ps = model.predict(X_sample)
+    if show:
+        plt.hist(ps, bins=bins)
+        plt.show()
+    begin, end = sms.DescrStatsW(ps.flatten()).tconfint_mean()
+    print('Confidence interval: ', begin, end)
+    print('Mean: ', np.mean([begin, end]))
+    print('True: ', y_test[sample_id])
+
```
Comment (gianlucatruda) [inline test.py:24]: It would be great if this could return a [named tuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) with the lower bound, mean, and upper bound. 

```python
Prediction = namedtuple('Prediction', ['lower', 'mean', 'upper'])
return Prediction(lower=<value>, mean=<value>, upper=<value>)
```

Then later we can write a scoring function that takes in `y_test` and evaluates the fraction of predictions that contain the ground truth in their confidence interval.

**Review by gianlucatruda — APPROVED** [review]
I pulled and ran the code in `run.py`. It works and starts training, but I quit because I don't have CUDA set up, so it's all on my CPU. 

We need to add Keras and Tensorflow dependencies to `requirements.txt`

I've also left some other minor comments. But, on the whole, great! 👍

**Comment by Serafim179** [comment]
> > HR as static feature
> 
> @Serafim179 does this mean you're using the ground truth value for the heart rate as an input feature for the model? That kinda violates the goals of the CoVital models, as we are trying to predict SPo2 based only on the video data. Or have I misunderstood something?

Yes, the ground truths are used for now, mainly as a placeholder for testing the performance.

**Comment by gianlucatruda** [comment]
> Yes, the ground truths are used for now, mainly as a placeholder for testing the performance.

Okay, we'll either need to update this when we do our evaluation, or justify this decision in the paper.
